### PR TITLE
[ESI] [Capnp] Compile library with exceptions

### LINKED ISSUE #3474: Cap'n Proto And Fibers / Exceptions
I did a bump of system-installed packages and there may be an issue with some distributions of Cap'n Proto for versions which have fibers enabled. It looks like fibers are incompatible with no-exception build options. E.g., I started to see the following build failure on a rebuild of CIRCT:

```
/nix/store/4w0pgdp0nv391sksvi179zjq3x7wvifz-capnproto-0.10.1/include/kj/async.h:39:6: error: "Fibers cannot be enabled when exceptions are disabled."
    #error "Fibers cannot be enabled when exceptions are disabled."
     ^
1 error generated.
```

This is probably a @teqdruid question on how to handle this.

I can always disable Cap'n Proto in my build.

### PR DESCRIPTION
**Pull request by teqdruid**
The EsiCosimCapnp library never gets linked into any MLIR code. It is
used exclusively to build the DPI shared library. As such, it can get
compiled with exceptions.

Should fix #3474 .

### CODE REVIEW

**Comment by teqdruid** [comment]
I'm not a cmake expert. I don't think this affects anything other than the libraries in the CMakeLists.txt. I checked with `ninja -v` and the other files seemed to have the `-fno-exceptions` flag whereas the failing file didn't.

Since I'm not able to repro the bug, do you mind testing this patch @dtzSiFive or @seldridge? I might need to add this magic elsewhere or it may not be effective.

**Comment by dtzSiFive** [comment]
Sure thing, I gave it a go!

This seems to be on the right track, fixes one of the errors!

Adding same snippet to `lib/Dialect/ESI/cosim/cosim_dpi_server/CMakeLists.txt` seems to fix the build.  I haven't investigated too closely, but it looks like EsiCosimDpiServer can have exceptions safely enabled as well (?).

I'm not sure about the replace, maybe we can just append `-fexceptions` instead?  Not sure if that approach will work for MSVC or what the best approach here is.  Anyway, poked at it a bit, this seems to work (sorry for rough delivery, have to run, but to show you what I mean):

```patch
From 83df552e6cbe65ac464a16cc811d3606320d14a3 Mon Sep 17 00:00:00 2001
From: Will Dietz <will.dietz@sifive.com>
Date: Wed, 20 Jul 2022 21:02:46 -0500
Subject: [PATCH] Also enable exceptions for EsiCosimDpiServer, append with
 try_compile_options?

---
 include/circt/Dialect/ESI/cosim/CMakeLists.txt       | 12 ++++++------
 .../ESI/cosim/cosim_dpi_server/CMakeLists.txt        |  7 +++++++
 2 files changed, 13 insertions(+), 6 deletions(-)

diff --git a/include/circt/Dialect/ESI/cosim/CMakeLists.txt b/include/circt/Dialect/ESI/cosim/CMakeLists.txt
index 2cd817093..bd871ff94 100644
--- a/include/circt/Dialect/ESI/cosim/CMakeLists.txt
+++ b/include/circt/Dialect/ESI/cosim/CMakeLists.txt
@@ -12,12 +12,6 @@ if(CapnProto_FOUND)
   set(COSIM_SCHEMA_HDR ${CIRCT_BINARY_DIR}/include/circt/Dialect/ESI/CosimSchema.h)
   configure_file(CosimSchema.h.in ${COSIM_SCHEMA_HDR})

-  if (MSVC)
-    string(REPLACE "/EHs-c-" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
-  else ()
-    string(REPLACE "-fno-exceptions" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
-  endif ()
-
   add_definitions(${CAPNP_DEFINITIONS})
   capnp_generate_cpp(COSIM_CAPNP_SRCS COSIM_CANPN_HDRS CosimDpi.capnp)
   add_library(EsiCosimCapnp
@@ -28,4 +22,10 @@ if(CapnProto_FOUND)
     PRIVATE
     CapnProto::capnp
     CapnProto::capnp-rpc)
+
+  if (MSVC)
+    string(REPLACE "/EHs-c-" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+  else ()
+    target_compile_options(EsiCosimCapnp PRIVATE -fexceptions)
+  endif ()
 endif()
diff --git a/lib/Dialect/ESI/cosim/cosim_dpi_server/CMakeLists.txt b/lib/Dialect/ESI/cosim/cosim_dpi_server/CMakeLists.txt
index 31f1b1471..14facfb4d 100644
--- a/lib/Dialect/ESI/cosim/cosim_dpi_server/CMakeLists.txt
+++ b/lib/Dialect/ESI/cosim/cosim_dpi_server/CMakeLists.txt
@@ -16,6 +16,13 @@ if(ESI_COSIM)
           RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib
 	  CXX_VISIBILITY_PRESET "default"
   )
+
+  if (MSVC)
+    string(REPLACE "/EHs-c-" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+  else ()
+    target_compile_options(EsiCosimDpiServer PRIVATE -fexceptions)
+  endif ()
+
   add_dependencies(EsiCosimDpiServer EsiCosimCapnp MtiPli)
   target_link_libraries(EsiCosimDpiServer PRIVATE
       CapnProto::kj CapnProto::kj-async CapnProto::kj-gzip
--
2.37.0
```

In short, it does seem however we manage to enable exceptions for these, that this does fix the build and seems worth pursuing.
Thanks for working on this!

**Comment by teqdruid** [comment]
> Adding same snippet to `lib/Dialect/ESI/cosim/cosim_dpi_server/CMakeLists.txt` seems to fix the build. I haven't investigated too closely, but it looks like EsiCosimDpiServer can have exceptions safely enabled as well (?).

Yes, it is. It's the other place to which I was referring.

> I'm not sure about the replace, maybe we can just append `-fexceptions` instead?

I don't like when things append stuff to a command line to negate something earlier in the command. Makes something which is already difficult to debug _much_ harder.

According to multiple posts, the correct way to write all the CXX flags these days is to write and use 'generators', which I (as usual with cmake) don't understand. Also, LLVM doesn't seem to use them so we'd have cmake files which are incompatible with LLVM's.

I can't think of a better way to remove that particular flag without that change being totally invasive in most of the other cmake files. The `string replace` method here is a hack but it is also the one exception, so I think it is warranted. (Also, I consider cmake in general to encourage hackiness so this is par for the course.) If anyone else knows a better way to remove that flag from the existing CXX_FLAGS variable, please let me know.

**Comment by dtzSiFive** [comment]
Sorry, yes this fixed the issue for me. I don't know CMake well enough/have bandwidth to suggest a better solution.  Although since we're the ones explicitly adding the `-fno-exceptions` at the top level we likely could handle this in a cleaner way without too much difficulty.  Maybe our CMake helpers are still borrowed from MLIR/LLVM and so it's not easy to accommodate such a change, I don't know.  LLVM's cmake has some support for opting in/out of exceptions, so we could likely do something like that.

Replacing is often considered dangerous as the command-line is entirely untyped/unknown string, and in theory we could mangle path names or otherwise botch a command-line.  As I think you know.  Anyway, I think that's rather unlikely here, as these strings are unlikely to occur in paths and at least for GCC-like's unlikely to have those as sub-strings of the flags (nothing like `-fno-exceptions-something`).

Trade-off against debugging pain and /maybe/ badness where appending the flag doesn't do what's expected although I'm not sure why that would happen.

I don't know generators well either.  FWIW LLVM does seem to use them in various places (looking for `'$<` in cmake-y things), but AFAIK not generally what downstream projects use.

Anyway glad it's fixed and sorry for not getting back to you sooner.  Something to fixup when we revisit our CMake, I suppose.

**Comment by teqdruid** [comment]
> Something to fixup when we revisit our CMake, I suppose.

Yep
